docs/en/operations/settings/settings.md

```diff
@@ -3144,6 +3144,17 @@ Result:
 └─────┴─────┴───────┘
 ```
 
+## enable_date32_results {#enable-date32-results}
+
+Enables or disables returning results of type Date32 for functions toStartOfYear, toStartOfISOYear, toStartOfQuarter, toStartOfMonth, toStartOfWeek, toMonday and toLastDayOfMonth.
+
+Possible values:
+
+-   0 — Functions return Date for all types of arguments.
+-   1 — Functions return Date32 for Date32 or DateTime64 arguments and Date otherwise.
+
+Default value: `0`.
+
```

Comment:
@rschu1ze Could you please discuss the better text for the NOTE box? My current variant:
After the last text of the NOTE ("the argument 2149-05-31 instead.")
Put additional text:

> For functions toStartOfYear, toStartOfISOYear, toStartOfQuarter, toStartOfMonth, toStartOfWeek, toMonday and toLastDayOfMonth may return type Date32 when option  enable_extended_results_for_datetime_functions = true.

Reply:
I thought about putting some text inside:

> The return type of toStartOf*, toLastDayOfMonth, toMonday functions described below is Date or DateTime. **NEW TEXT ABOUT NEW OPTION**. Though these functions can take values of the extended types Date32 and DateTime64 as an argument

But probably it's not a good idea. It will complicate further reading.

Reply:
As a proposal, let's replace the NOTE box by the following text. (I omitted the formatting).
```
The return type of the toStartOf*, toLastDayOfMonth, toMonday functions described below is determined by the configuration parameter "enable_date32_results" which is "false" by default.

Behavior for
- "enable_date32_results = false": Functions toStartOf*, toLastDayOfMonth, toMonday return Date or DateTime. Though these functions can take values of the extended types Date32 and DateTime64 as an argument, passing them a time outside the normal range (year 1970 to 2149 for Date / 2106 for DateTime) will produce wrong results. In case argument is out of normal range:
    - If the argument is smaller than 1970, the result will be calculated from the argument 1970-01-01 (00:00:00) instead.
    - If the return type is DateTime and the argument is larger than 2106-02-07 08:28:15, the result will be calculated from the argument 2106-02-07 08:28:15 instead.
    - If the return type is Date and the argument is larger than 2149-06-06, the result will be calculated from the argument 2149-06-06 instead.
    - If toLastDayOfMonth is called with an argument greater then 2149-05-31, the result will be calculated from the argument 2149-05-31 instead.
- "enable_date32_results = true": Functions toStartOf*, toLastDayOfMonth, toMonday return Date or DateTime if their argument is a Date or DateTime, and they return Date32 or DateTime64 if their argument is a Date32 or DateTime64.
```

Note that all for the second case, all the special cases are generally removed from the description.

EDIT: Thinking about it, the last special case (toLastDayOfMonth with a Date argument > 2149-05-31) will need special treatment even with your new approach. In some other PR, I suggested to throw an exception for this edge case because the result cannot be represented as a Date type and we could remove the ugly clamping. But a brief note about this special case could also be made not in the NOTE box but in the documentation for function "toLastDayOfMonth".

Reply:
> "enable_date32_results = true": Functions toStartOf*, toLastDayOfMonth, toMonday return Date or DateTime if their argument is a Date or DateTime, and they return Date32 or DateTime64 if their argument is a Date32 or DateTime64.

OK. One note: not all functions toStartOf* may return Date32. Because some functions return DateTime for example, toStartOfHour. Maybe it's better to list all functions:

"enable_date32_results = true": Functions toStartOfYear, toStartOfISOYear, toStartOfQuarter, toStartOfMonth, toStartOfWeek, toLastDayOfMonth, toMonday return Date or DateTime if

Reply:
Yepp, that makes sense.

Reply:
Done

Reply:
@rschu1ze it seems like I resolved all the questions from this list and I pushed the changes. The one which is remaining is

> EDIT: Thinking about it, the last special case (toLastDayOfMonth with a Date argument > 2149-05-31) will need special treatment even with your new approach. In some other PR, I suggested to throw an exception for this edge case because the result cannot be represented as a Date type and we could remove the ugly clamping. But a brief note about this special case could also be made not in the NOTE box but in the documentation for function "toLastDayOfMonth".

Yes, you correctly said that current PR does not solve 2149-05-31 problem and it should be solved somehow in future PRs then we remove all legacy code. Exception is one of possible solutions. But I am not sure if it's a "ClickHouse-way" for such situation. For example, a client processes a big array of data and in some raws there is 2149-06-02 it will raise an exception and it will break all processing. While "2149-05-31-way" maybe is better while it's not ideal. 

Right now what kind of note should be added for toLastDayOfMonth function?

Reply:
I am too new in the team to know all of ClickHouse's philosophy. What I have seen a lot is that the default behavior is usually the intuitive one but that it can also be modified for special needs using config parameters. In either case, function toLastDayOfMonth() would need an `if` to check. Since we talk about an edge case, we may use `likely()/unlikely()` on that `if` and there will be little performance impact. So what about a nested `if` where the actual behavior can be tuned by a cfg parameter between narrowing and an exception? In any case, like you said it's a trade-off and overall there is no ideal solution which works for all use cases.

Reply:
I also too new to ClickHouse (I have been working with ClickHouse since the last six months). My first surprise was, for example:
`select toDate('1950-01-01');`
It returns `1970-01-01` 
I remember being discussed it with @ilejn that the normal behavior is the raise an exception. But then I adopted to think that "cropping/clamping" is the ClickHouse way, while I am not sure if I am right in thinking so.
`likely()/unlikely() ` in execute() may help. But I think it's can damage the performance because execute()  is called on any value calculation. So some study/experiments are required on this subject.

Reply:
Okay. About the note .. what about the following as docs of toLastDayOfMonth? (it's more or less copypasta from what we had previously in the NOTES box).

```
Rounds up a date or date with time to the last day of the month. Returns a date.

If toLastDayOfMonth is called with an argument of type Date greater then 2149-05-31, the result will be calculated from the argument 2149-05-31 instead.
```

Reply:
Done. Added this note.